python/lsst/ts/externalscripts/base_make_calibrations.py

```diff
@@ -838,21 +838,145 @@ async def count_failed_verification_tests(
+                )
+
+        # verify_stats
+        final_report_string += "Test types that failed verification per exposure,\n|"
+        final_report_string += "detector, and amplifier:\n|"
+
+        for exposure in [key for key in verify_stats if key != "SUCCESS"]:
+            final_report_string += f"\t|Exposure ID: {exposure}\n|"
+            if "FAILURES" in verify_stats[exposure]:
+                # det name | amp | test type
+                for info in verify_stats[exposure]["FAILURES"]:
+                    final_report_string += f"\t| \t| {info}\n|"
+            else:
+                final_report_string += (
+                    "No failures in 'verify_stats' for this exposure."
+                )
+
+        # thresholds_report
+        final_report_string += "Threshold values:\n|"
+        final_report_string += (
+            "\t| Acceptable maximum number of failures per detector per test type:\n|"
+        )
+        final_report_string += f"{thresholds_report['MAX_FAILURES_PER_DETECTOR_PER_TEST_TYPE_THRESHOLD']}\n|"
+
+        final_report_string += (
+            "This value is controlled by the configuration parameter\n|"
+        )
+        final_report_string += "'number_verification_tests_threshold_IMGTYPE'\n|"
+
+        final_report_string += "\t|Acceptable maximum number of failed detectors:]\n|"
+        final_report_string += (
+            f"{thresholds_report['MAX_FAILED_DETECTORS_THRESHOLD']}\n|"
+        )
+
+        final_report_string += (
+            "\t|Acceptable maximum number of failed tests per exposure:\n|"
+        )
+        final_report_string += (
+            f"{thresholds_report['MAX_FAILED_TESTS_PER_EXPOSURE_THRESHOLD']}\n|"
+        )
+
+        final_report_string += "\t|Acceptable maximum number of failed exposures:\n|"
+        final_report_string += (
+            f"{thresholds_report['MAX_FAILED_EXPOSURES_THRESHOLD']}\n|"
+        )
+        final_report_string += (
+            "\t|Final number of exposures that failed verification:\n|"
+        )
+        final_report_string += (
+            f"{thresholds_report['FINAL_NUMBER_OF_FAILED_EXPOSURES']}\n|"
+        )
+
+        final_report_string += (
+            "Verification failure criterium: if, for at least une type of test,\n|"
+        )
+        final_report_string += (
+            "the majority of tests fail in the majority of detectors and the\n|"
+        )
+        final_report_string += (
+            "the majority of exposures, verification will fail and the calibration\n|"
+        )
+        final_report_string += "will not be certified. "
+
+        final_report_string += (
+            "In terms of the threshold values, this amounts for the condition that\n|"
+        )
+        final_report_string += (
+            "the final number of exposures that failed verification is greater than\n|"
+        )
+        final_report_string += (
+            "or equal to the acceptable maximum number of failed exposures"
```

Comment:
Creating a string by repeatedly appending to it is very inefficient.

Please consider accumulating a list of entries and useing join, instead. If `"\n|"` appears before every entry then you can omit it and use it in the join, instead. Here is an example:
```
final_report_strings = [
    "Exposures with verification tests that failed:",
    " ".join(str(exposure) for exposure in verify_report),
    "Number of tests that failed per test type:"
    f"\t| Exposure ID: {exposure}",
]
for test_type in verify_report[exposure]:
    final_report_strings.append(
        f"\t|{test_type}: {verify_report[exposure][test_type]}"
    )
final_report_strings += [
    "Test types that failed verification per exposure,",
    "detector, and amplifier:",
    ...
]
...
final_report_string = "\n|".join(final_report_strings)
```

And any "line" can be as many lines of code as you want, e.g. if the list of exposures should appear right after "Exposures with..." then just delete the comma separating those two entries to concatenate those two lines, or add a `+` to make it a bit more obvious that the concatenation is intentional:
```
final_report_strings = [
    "Exposures with verification tests that failed: "
    + " ".join(str(exposure) for exposure in verify_report),
```

Reply:
I'll be removing the "|"; we determined with Tiago that they don't work.